### PR TITLE
Integrate cppcheck with travis-ci

### PR DESCRIPTION
**Pull request by manish3kumar**
Signed-off-by: Manish, Kumar <manish3.kumar@intel.com>

### CODE REVIEW

**Comment by manish3kumar** [comment]
Hartwork,

please review and help to merge the change as it integrates the cppcheck tool for the static analysis.
I see that build is failed https://travis-ci.org/libexpat/libexpat/builds/557250095 due to issues found due to my patch.

33.66s$ cppcheck --quiet --error-exitcode=1 .
[expat/lib/xmltok.c:1506] -> [expat/lib/xmltok.c:1506]: (warning) The address of local variable 'latin1_encoding' might be accessed at non-zero index.
[expat/tests/runtests.c:647]: (error) Address of local auto-variable assigned to a function parameter.
[expat/tests/runtests.c:4976]: (warning) Possible null pointer dereference: text
[expat/xmlwf/xmlwin32url.cxx:54]: (error) There is an unknown macro here somewhere. Configuration is required. If STDMETHODIMP_ is a macro then please configure it.
(information) Too many #ifdef configurations - cppcheck only checks 12 configurations. Use --force to check all configurations. For more details, use --enable=information.
The command "cppcheck --quiet --error-exitcode=1 ." exited with 1.

Thanks,
Manish

**Comment by manish3kumar** [comment]
If you merge the change, then it's good for all future patches which gets integrated into libexpat. 

Please let me know your review comments if there are any.

File: .travis.yml [hunk 1]
```diff
@@ -12,6 +12,10 @@ dist: trusty
 # https://github.com/travis-ci/travis-ci/issues/9033
 sudo: required
 
+before_install:
+  - if [ "$TRAVIS_OS_NAME" = "osx" ]; then brew update          ; fi
```
Comment (hartwork) [inline .travis.yml]: This seems to duplicate the `addons.homebrew.update = True` below

File: .travis.yml [hunk 2]
```diff
@@ -12,6 +12,10 @@ dist: trusty
...
+  - if [ "$TRAVIS_OS_NAME" = "osx" ]; then brew install cppcheck; fi
```
Comment (hartwork) [inline .travis.yml]: Please add to `Brewfile` in the root folder, instead.

File: .travis.yml [hunk 3]
```diff
@@ -44,6 +48,9 @@ addons:
       - dos2unix
       - lzip
       - rpm
+      - cppcheck
 
 script:
   - ./.travis.sh
+  # cppcheck
+  - cppcheck --quiet --error-exitcode=1 .
```
Comment (hartwork) [inline .travis.yml]: Please use a new `MODE=cppcheck` instead and call `cppcheck` from `.travis.sh`, instead. This way, we run it once for every platform, not more.

**Comment by hartwork** [comment]
Please rebase against latest master — all of #282 #283 #284 #285 have been fixed.

File: .travis.yml (hunk 1 above)
Comment (manish3kumar) [reply .travis.yml]: Done, removed it, because it's taken care into the addons.homebrew.update = true.

File: .travis.yml (hunk 2 above)
Comment (manish3kumar) [reply .travis.yml]: This is already taken care.

File: .travis.yml (hunk 3 above)
Comment (manish3kumar) [reply .travis.yml]: I added new mode, but I am not sure about calling them from .travis.sh file for all OS/platform. Could you please provide more details?

File: .travis.yml (hunk 2 above)
Comment (hartwork) [reply .travis.yml]: How is it taken care of? [Brewfile](https://github.com/libexpat/libexpat/blob/master/Brewfile) does not contain an entry for cppcheck yet, in this pull request.

File: .travis.yml (hunk 3 above)
Comment (hartwork) [reply .travis.yml]: Travis covers macOS and Linux for us so need to worry about Windows — that's AppVeyor.
Calling `cppcheck` from `.travis.sh` should work for both operating systems.
I was thinking of a new block similar to `elif [[ ${MODE} = cppcheck ]]; then`.
Does that answer the question?

File: .travis.yml (hunk 2 above)
Comment (manish3kumar) [reply .travis.yml]: I will upload the change today,  once the cppcheck mode check is taken care by me.

File: .travis.yml (hunk 3 above)
Comment (manish3kumar) [reply .travis.yml]: Yes, I was also thinking the same, but not sure how will it be called.

Added mode cppcheck below.

if [[ ${MODE} = distcheck ]]; then
    ./configure ${CONFIGURE_ARGS}
    make distcheck

    mkdir -p ~/rpmbuild/{BUILD,BUILDROOT,RPMS,SOURCES,SPECS,SRPMS}
    ln -v -s "$PWD"/expat-*.tar.bz2 ~/rpmbuild/SOURCES/
    rpmbuild -ba expat.spec
elif [[ ${MODE} = cmake-oos ]]; then
    mkdir build
    cd build
    cmake ${CMAKE_ARGS} ..
    make all test
    make DESTDIR="${PWD}"/ROOT install
    find ROOT -printf "%P\n" | sort
elif [[ ${MODE} = cppcheck ]]; then
    cppcheck --quiet --error-exitcode=1 .
else
    ./qa.sh
fi

I am thinking to call it through .travis.yml file 

matrix:
  include:
    - os: linux
      env: MODE=cmake-oos CMAKE_ARGS=-DXML_ATTR_INFO=ON cppcheck
    - os: linux
      env: MODE=qa-sh QA_COMPILER=clang CC=clang CXX=clang++ LD=clang++ QA_SANITIZER=address CONFIGURE_ARGS=--enable-xml-attr-info cppcheck
    - os: linux
      env: MODE=qa-sh QA_COMPILER=clang CC=clang CXX=clang++ LD=clang++ QA_SANITIZER=memory cppcheck
    - os: linux
      env: MODE=qa-sh QA_COMPILER=clang CC=clang CXX=clang++ LD=clang++ QA_SANITIZER=undefined cppcheck
    - os: linux
      env: MODE=qa-sh QA_COMPILER=gcc   CC=gcc   CXX=g++     LD=ld      QA_PROCESSOR=gcov cppcheck

is this enough? or any other thoughts. Please let me know your thoughts.

Comment (hartwork) [reply .travis.yml]: This part looks good:
```
elif [[ ${MODE} = cppcheck ]]; then
    cppcheck --quiet --error-exitcode=1 .
```
Please check in more detail how Travis matrix env works and make a new proposal.

Thanks!

Comment (hartwork) [reply .travis.yml]: To be more concrete, a plain addition of ...
```
    - os: linux
      env: MODE=cppcheck
```
... to the matrix, would be what I'd expect — nothing more.

**Comment by manish3kumar** [comment]
https://github.com/libexpat/libexpat/pull/290 -- Started new pull request.

File: .travis.yml (hunk 2 above)
Comment (manish3kumar) [reply .travis.yml]: Please check new pull request https://github.com/libexpat/libexpat/pull/290

File: .travis.yml (hunk 3 above)
Comment (manish3kumar) [reply .travis.yml]: Please check new pull request https://github.com/libexpat/libexpat/pull/290

**Comment by manish3kumar** [comment]
Please check new pull request https://github.com/libexpat/libexpat/pull/290

**Comment by manish3kumar** [comment]
Please check new pull request https://github.com/libexpat/libexpat/pull/290

**Pull request closed without merging**